lib/src/analysis_options/merge_options.dart

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// Merges a [defaults] options set with an [overrides] options set using
+/// simple override semantics, suitable for merging two configurations where
+/// one defines default values that are added to (and possibly overridden) by an
+/// overriding one.
+///
+/// The merge rules are:
+///
+/// *   Lists are concatenated without duplicates.
+/// *   A list of strings is promoted to a map of strings to `true` when merged
+///     with another map of strings to booleans. For example `['opt1', 'opt2']`
+///     is promoted to `{'opt1': true, 'opt2': true}`.
+/// *   Maps unioned. When both have the same key, the corresponding values are
+///     merged, recursively.
+/// *   Otherwise, a non-`null` override replaces a default value.
+Object? merge(Object? defaults, Object? overrides) {
+  return switch ((defaults, overrides)) {
+    (List(isAllStrings: true) && var list, Map(isToBools: true)) =>
```

Comment:
Wouldn't it be nice if you could just write:
```dart
  (List(every(isA<String>): true) && var list, Map(values.every(isA<bool>))) =>
```
and not need extension getters, just a simple `bool isA<T>(Object? value) => value is T;`. :wink: 
and not *need* a getter? :wink:

Reply:
Yes! I would love some kind of way to embed predicates inside patterns.

Reply:
Well, do go vote on https://github.com/dart-lang/language/issues/2433 then! :)